// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: usercenter/v1/secret.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Secret with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Secret) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Secret with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in SecretMultiError, or nil if none found.
func (m *Secret) ValidateAll() error {
	return m.validate(true)
}

func (m *Secret) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserID

	// no validation rules for Name

	// no validation rules for SecretID

	// no validation rules for SecretKey

	// no validation rules for Expires

	// no validation rules for Status

	// no validation rules for Description

	if all {
		switch v := interface{}(m.GetCreatedAt()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SecretValidationError{
					field:  "CreatedAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SecretValidationError{
					field:  "CreatedAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetCreatedAt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SecretValidationError{
				field:  "CreatedAt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetUpdatedAt()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SecretValidationError{
					field:  "UpdatedAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SecretValidationError{
					field:  "UpdatedAt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetUpdatedAt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SecretValidationError{
				field:  "UpdatedAt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return SecretMultiError(errors)
	}

	return nil
}

// SecretMultiError is an error wrapping multiple validation errors returned by
// Secret.ValidateAll() if the designated constraints aren't met.
type SecretMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SecretMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SecretMultiError) AllErrors() []error { return m }

// SecretValidationError is the validation error returned by Secret.Validate if
// the designated constraints aren't met.
type SecretValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SecretValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SecretValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SecretValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SecretValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SecretValidationError) ErrorName() string { return "SecretValidationError" }

// Error satisfies the builtin error interface
func (e SecretValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSecret.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SecretValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SecretValidationError{}

// Validate checks the field values on CreateSecretRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateSecretRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateSecretRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateSecretRequestMultiError, or nil if none found.
func (m *CreateSecretRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateSecretRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	// no validation rules for Expires

	// no validation rules for Description

	if len(errors) > 0 {
		return CreateSecretRequestMultiError(errors)
	}

	return nil
}

// CreateSecretRequestMultiError is an error wrapping multiple validation
// errors returned by CreateSecretRequest.ValidateAll() if the designated
// constraints aren't met.
type CreateSecretRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateSecretRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateSecretRequestMultiError) AllErrors() []error { return m }

// CreateSecretRequestValidationError is the validation error returned by
// CreateSecretRequest.Validate if the designated constraints aren't met.
type CreateSecretRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateSecretRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateSecretRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateSecretRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateSecretRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateSecretRequestValidationError) ErrorName() string {
	return "CreateSecretRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreateSecretRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateSecretRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateSecretRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateSecretRequestValidationError{}

// Validate checks the field values on CreateSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateSecretResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateSecretResponseMultiError, or nil if none found.
func (m *CreateSecretResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateSecretResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for SecretID

	if len(errors) > 0 {
		return CreateSecretResponseMultiError(errors)
	}

	return nil
}

// CreateSecretResponseMultiError is an error wrapping multiple validation
// errors returned by CreateSecretResponse.ValidateAll() if the designated
// constraints aren't met.
type CreateSecretResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateSecretResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateSecretResponseMultiError) AllErrors() []error { return m }

// CreateSecretResponseValidationError is the validation error returned by
// CreateSecretResponse.Validate if the designated constraints aren't met.
type CreateSecretResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateSecretResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateSecretResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateSecretResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateSecretResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateSecretResponseValidationError) ErrorName() string {
	return "CreateSecretResponseValidationError"
}

// Error satisfies the builtin error interface
func (e CreateSecretResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateSecretResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateSecretResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateSecretResponseValidationError{}

// Validate checks the field values on UpdateSecretRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateSecretRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateSecretRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateSecretRequestMultiError, or nil if none found.
func (m *UpdateSecretRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateSecretRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	if m.Expires != nil {
		// no validation rules for Expires
	}

	if m.Status != nil {
		// no validation rules for Status
	}

	if m.Description != nil {
		// no validation rules for Description
	}

	if len(errors) > 0 {
		return UpdateSecretRequestMultiError(errors)
	}

	return nil
}

// UpdateSecretRequestMultiError is an error wrapping multiple validation
// errors returned by UpdateSecretRequest.ValidateAll() if the designated
// constraints aren't met.
type UpdateSecretRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateSecretRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateSecretRequestMultiError) AllErrors() []error { return m }

// UpdateSecretRequestValidationError is the validation error returned by
// UpdateSecretRequest.Validate if the designated constraints aren't met.
type UpdateSecretRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateSecretRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateSecretRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateSecretRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateSecretRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateSecretRequestValidationError) ErrorName() string {
	return "UpdateSecretRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateSecretRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateSecretRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateSecretRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateSecretRequestValidationError{}

// Validate checks the field values on UpdateSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateSecretResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateSecretResponseMultiError, or nil if none found.
func (m *UpdateSecretResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateSecretResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return UpdateSecretResponseMultiError(errors)
	}

	return nil
}

// UpdateSecretResponseMultiError is an error wrapping multiple validation
// errors returned by UpdateSecretResponse.ValidateAll() if the designated
// constraints aren't met.
type UpdateSecretResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateSecretResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateSecretResponseMultiError) AllErrors() []error { return m }

// UpdateSecretResponseValidationError is the validation error returned by
// UpdateSecretResponse.Validate if the designated constraints aren't met.
type UpdateSecretResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateSecretResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateSecretResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateSecretResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateSecretResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateSecretResponseValidationError) ErrorName() string {
	return "UpdateSecretResponseValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateSecretResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateSecretResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateSecretResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateSecretResponseValidationError{}

// Validate checks the field values on DeleteSecretRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteSecretRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteSecretRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteSecretRequestMultiError, or nil if none found.
func (m *DeleteSecretRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteSecretRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	if len(errors) > 0 {
		return DeleteSecretRequestMultiError(errors)
	}

	return nil
}

// DeleteSecretRequestMultiError is an error wrapping multiple validation
// errors returned by DeleteSecretRequest.ValidateAll() if the designated
// constraints aren't met.
type DeleteSecretRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteSecretRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteSecretRequestMultiError) AllErrors() []error { return m }

// DeleteSecretRequestValidationError is the validation error returned by
// DeleteSecretRequest.Validate if the designated constraints aren't met.
type DeleteSecretRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteSecretRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteSecretRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteSecretRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteSecretRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteSecretRequestValidationError) ErrorName() string {
	return "DeleteSecretRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteSecretRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteSecretRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteSecretRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteSecretRequestValidationError{}

// Validate checks the field values on DeleteSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteSecretResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteSecretResponseMultiError, or nil if none found.
func (m *DeleteSecretResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteSecretResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return DeleteSecretResponseMultiError(errors)
	}

	return nil
}

// DeleteSecretResponseMultiError is an error wrapping multiple validation
// errors returned by DeleteSecretResponse.ValidateAll() if the designated
// constraints aren't met.
type DeleteSecretResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteSecretResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteSecretResponseMultiError) AllErrors() []error { return m }

// DeleteSecretResponseValidationError is the validation error returned by
// DeleteSecretResponse.Validate if the designated constraints aren't met.
type DeleteSecretResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteSecretResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteSecretResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteSecretResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteSecretResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteSecretResponseValidationError) ErrorName() string {
	return "DeleteSecretResponseValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteSecretResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteSecretResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteSecretResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteSecretResponseValidationError{}

// Validate checks the field values on GetSecretRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetSecretRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetSecretRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetSecretRequestMultiError, or nil if none found.
func (m *GetSecretRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetSecretRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	if len(errors) > 0 {
		return GetSecretRequestMultiError(errors)
	}

	return nil
}

// GetSecretRequestMultiError is an error wrapping multiple validation errors
// returned by GetSecretRequest.ValidateAll() if the designated constraints
// aren't met.
type GetSecretRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetSecretRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetSecretRequestMultiError) AllErrors() []error { return m }

// GetSecretRequestValidationError is the validation error returned by
// GetSecretRequest.Validate if the designated constraints aren't met.
type GetSecretRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetSecretRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetSecretRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetSecretRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetSecretRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetSecretRequestValidationError) ErrorName() string { return "GetSecretRequestValidationError" }

// Error satisfies the builtin error interface
func (e GetSecretRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetSecretRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetSecretRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetSecretRequestValidationError{}

// Validate checks the field values on GetSecretResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetSecretResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetSecretResponseMultiError, or nil if none found.
func (m *GetSecretResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetSecretResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetSecret()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetSecretResponseValidationError{
					field:  "Secret",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetSecretResponseValidationError{
					field:  "Secret",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetSecret()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetSecretResponseValidationError{
				field:  "Secret",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetSecretResponseMultiError(errors)
	}

	return nil
}

// GetSecretResponseMultiError is an error wrapping multiple validation errors
// returned by GetSecretResponse.ValidateAll() if the designated constraints
// aren't met.
type GetSecretResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetSecretResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetSecretResponseMultiError) AllErrors() []error { return m }

// GetSecretResponseValidationError is the validation error returned by
// GetSecretResponse.Validate if the designated constraints aren't met.
type GetSecretResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetSecretResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetSecretResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetSecretResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetSecretResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetSecretResponseValidationError) ErrorName() string {
	return "GetSecretResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetSecretResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetSecretResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetSecretResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetSecretResponseValidationError{}

// Validate checks the field values on ListSecretRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListSecretRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListSecretRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListSecretRequestMultiError, or nil if none found.
func (m *ListSecretRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListSecretRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Offset

	// no validation rules for Limit

	if len(errors) > 0 {
		return ListSecretRequestMultiError(errors)
	}

	return nil
}

// ListSecretRequestMultiError is an error wrapping multiple validation errors
// returned by ListSecretRequest.ValidateAll() if the designated constraints
// aren't met.
type ListSecretRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListSecretRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListSecretRequestMultiError) AllErrors() []error { return m }

// ListSecretRequestValidationError is the validation error returned by
// ListSecretRequest.Validate if the designated constraints aren't met.
type ListSecretRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListSecretRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListSecretRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListSecretRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListSecretRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListSecretRequestValidationError) ErrorName() string {
	return "ListSecretRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListSecretRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListSecretRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListSecretRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListSecretRequestValidationError{}

// Validate checks the field values on ListSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListSecretResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListSecretResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListSecretResponseMultiError, or nil if none found.
func (m *ListSecretResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListSecretResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Total

	for idx, item := range m.GetSecrets() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListSecretResponseValidationError{
						field:  fmt.Sprintf("Secrets[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListSecretResponseValidationError{
						field:  fmt.Sprintf("Secrets[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListSecretResponseValidationError{
					field:  fmt.Sprintf("Secrets[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListSecretResponseMultiError(errors)
	}

	return nil
}

// ListSecretResponseMultiError is an error wrapping multiple validation errors
// returned by ListSecretResponse.ValidateAll() if the designated constraints
// aren't met.
type ListSecretResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListSecretResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListSecretResponseMultiError) AllErrors() []error { return m }

// ListSecretResponseValidationError is the validation error returned by
// ListSecretResponse.Validate if the designated constraints aren't met.
type ListSecretResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListSecretResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListSecretResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListSecretResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListSecretResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListSecretResponseValidationError) ErrorName() string {
	return "ListSecretResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListSecretResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListSecretResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListSecretResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListSecretResponseValidationError{}
